server/storages/snowflake_suspender.go

```diff
@@ -0,0 +1,196 @@
+			Params:    config.Parameters,
+		}
+		connectionString, err := sf.DSN(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		dataSource, err := sql.Open("snowflake", connectionString)
+		if err != nil {
+			return nil, err
+		}
+		suspender = &SnowflakeSuspender{key: key, warehouse: config.Warehouse,
+			lastRequestTime:   time.Now(),
+			keepAwakeDuration: keepAwakeDuration,
+			dataSource:        dataSource,
+			monitorKeeper:     monitorKeeper,
+			stopped: 		   make(chan bool),
+		}
+		suspender.start()
+		suspenders[key] = suspender
+		usageCounters[key]++
+	}
+	return suspender, nil
+}
+
+func ReleaseSnowflakeSuspender(suspender *SnowflakeSuspender) {
+	globalMutex.Lock()
+	defer globalMutex.Unlock()
+	var usg = usageCounters[suspender.key]
+	switch {
+	case usg == 1:
+		delete(suspenders, suspender.key)
+		suspender.stop()
+		usageCounters[suspender.key] = 0
+	case usg > 1:
+		usageCounters[suspender.key]--
+	default:
+		logging.Errorf("Trying to release orphaned SnowflakeSuspender: %v", suspender.key)
+	}
+}
+
+func (awk *SnowflakeSuspender) start() {
+	logging.Infof("Starting SnowflakeSuspender %s", awk.key)
+	awk.ticker = time.NewTicker(time.Second)
+	safego.RunWithRestart(func() {
+		for {
+			select {
+			case <-awk.ticker.C:
+				awk.sync()
+			case <-awk.stopped:
+				awk.ticker.Stop()
+				return
+			}
+		}
+	})
+}
+
+func (awk *SnowflakeSuspender) stop() {
+	logging.Infof("Stopping SnowflakeSuspender %s", awk.key)
+	close(awk.stopped)
+	err := awk.dataSource.Close()
+	if err != nil {
+		logging.Errorf("failed to close SnowflakeSuspender's datasource: %v", err)
+	}
+}
+
+func (awk *SnowflakeSuspender) sync() {
+	awk.Lock()
+	defer awk.Unlock()
+	logging.Debugf("Sync %v", awk)
+
+	if awk.lastRequestTime.After(awk.lastRequestTimeInStorage) {
+
+		//since we run every second, and we don't require to the second precision
+		//it is not critical if we overwrite more fresh value from other instance
+		err := awk.monitorKeeper.UpdateLastRunTime(awk.key, awk.lastRequestTime)
+		if err != nil {
+			logging.Errorf("failed to update last sql execution time for: %v err: %v", awk.key, err)
+		}
+		awk.lastRequestTimeInStorage = awk.lastRequestTime
+	} else if !awk.suspended && awk.runningTasksCount == 0 &&
+		time.Since(awk.lastRequestTime) > awk.keepAwakeDuration {
+		//check maybe other instances run sql and refreshed 'last run time'
+		tm, err := awk.monitorKeeper.GetLastRunTime(awk.key)
+		if err != nil {
+			logging.Errorf("failed to get last sql execution time for: %v err: %v", awk.key, err)
+		}
+		if tm.Sub(awk.lastRequestTime).Seconds() >= 1 {
+			logging.Debugf("Sync %v fresh in storage: %v", awk.key, tm)
+			awk.lastRequestTimeInStorage = tm
+			awk.lastRequestTime = tm
+		} else {
+			logging.Infof("%v Suspending warehouse", awk.key)
+			_, err := awk.dataSource.Exec(fmt.Sprintf(alterQuery, awk.warehouse))
+			if err != nil && !strings.HasSuffix(err.Error(), fmt.Sprintf(alreadySuspendedMsg, awk.warehouse)) {
```

Comment:
Since Snowflake returns the uppercased warehouse name in error, this condition seems to be unreliable. Try to use error codes instead. (like in CreateSnowflakeAdapter() func in storages package)